library/ccm.c

```diff
@@ -307,15 +388,33 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
                 break;
     }
 
+    if( len_left > 0 )
+    {
+        return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+    }
     /*
      * Authentication: reset counter and crypt/mask internal tag
      */
+    flow_ctrl_local = 0;
     for( i = 0; i < q; i++ )
+    {
+        flow_ctrl_local++;
         ctr[15-i] = 0;
+    }
+    CHECK_FLOW_CTRL( q, fc6 );
 
     CTR_CRYPT( y, y, 16 );
+    CHECK_FLOW_CTRL( 16, fc7 );
```

Comment:
@AndrzejKurek , why do we need to check flow control in different macro? 
Is it less secure to implement something like :
 #define UPDATE_CBC_MAC                                          \
flow_ctrl_local =0;                                                        \
    for( i = 0; i < 16; i++ )                                               \
    {                                                                       \
        flow_ctrl_local++;                                            \
        y[i] ^= b[i];                                                      \
        y[i] ^= b[i];                                                     \
    }                                                                       \
                                                                          
 if  ( flow_ctrl_local !=16) return MBEDTLS_ERR_PLATFORM_FAULT_DETECTED; \	                                                                         
   \
    if( ( ret = mbedtls_cipher_update( &ctx->cipher_ctx, y, 16, y, &olen ) ) != 0 ) \	    if( ( ret = mbedtls_cipher_update( &ctx->cipher_ctx, y, 16, y, &olen ) ) != 0 ) \
        return( ret );	        return( ret );

Reply:
I separated the flow control checking macro so that it can be used without `UPDATE_CBC_MAC` or `CTR_CRYPT`, for example in line 404.

Reply:
@AndrzejKurek , I see, it just seems to me shorter to add one line to CTR_CRYPT and one line to UPDATE_CBC_MAC , than use 5 lines macro . Let's talk to understand flow_flag usage in this new macro. I am worried about code size we added to ccm module

Reply:
There will be more lines, as we have to set the appropriate flag if everything is OK. This won't change the codesize much (if at all), and will make us remove the additional macro, but more code will have to be added instead in lines like 404.

Reply:
The `flow_ctrl_local`  changes inside `UPDATE_CBC_MAC` and then it is verified inside `CHECK_FLOW_CTRL`, but this is not visible without more detailed code analysis. Could you use flow_ctrl_local for these macros a an argument?